controller/src/beerocks/bml/internal/bml_internal.cpp

```diff
@@ -1767,6 +1839,154 @@ int bml_internal::start_dcs_single_scan(const sMacAddr &mac, int dwell_time_ms,
+
+    // If the socket is not valid, attempt to re-establish the connection
+    if (!m_sockMaster) {
+        int iRet = connect_to_master();
+        if (iRet != BML_RET_OK) {
+            LOG(ERROR) << " Unable to create context, connect_to_master failed!";
+            return iRet;
+        }
+    }
+
+    // Initialize the promise for receiving the response
+    beerocks::promise<bool> prmClientListGet;
+    m_prmClientListGet = &prmClientListGet;
+    int iOpTimeout     = RESPONSE_TIMEOUT; // Default timeout
+    m_client_list      = &client_list;
+    m_client_list_size = client_list_size;
+
+    auto request = message_com::create_vs_message<
+        beerocks_message::cACTION_BML_CLIENT_GET_CLIENT_LIST_REQUEST>(cmdu_tx);
+
+    if (!request) {
+        LOG(ERROR) << "Failed building cACTION_BML_CLIENT_GET_CLIENT_LIST_REQUEST message!";
+        return (-BML_RET_OP_FAILED);
+    }
+
+    // Build and send the message
+    if (!message_com::send_cmdu(m_sockMaster, cmdu_tx)) {
+        LOG(ERROR) << "Failed sending param get message!";
+        m_prmClientListGet = nullptr;
+        m_client_list      = nullptr;
+        return (-BML_RET_OP_FAILED);
+    }
+    LOG(DEBUG) << "cACTION_BML_CLIENT_GET_CLIENT_LIST_REQUEST sent";
+
+    int iRet = BML_RET_OK;
+
+    if (!m_prmClientListGet->wait_for(iOpTimeout)) {
+        LOG(WARNING) << "Timeout while waiting for client list get response...";
+        iRet = -BML_RET_TIMEOUT;
+    }
+
+    // Clear the scan results members
+    m_client_list = nullptr;
+
+    // Clear the promise holder
+    m_prmClientListGet = nullptr;
+
+    LOG_IF(iRet != BML_RET_OK, ERROR) << "Get client list returned with error code:" << iRet;
+
+    return BML_RET_OK;
+}
+
+int bml_internal::client_set_client(const sMacAddr &sta_mac, const BML_CLIENT_CONFIG &client_config)
+{
+    LOG(DEBUG) << "client_set_client";
+
+    auto request =
+        message_com::create_vs_message<beerocks_message::cACTION_BML_CLIENT_SET_CLIENT_REQUEST>(
+            cmdu_tx);
+
+    if (!request) {
+        LOG(ERROR) << "Failed building cACTION_BML_CLIENT_SET_CLIENT_REQUEST message!";
+        return (-BML_RET_OP_FAILED);
+    }
+
+    request->sta_mac()                               = sta_mac;
+    request->client_config().stay_on_initial_radio   = client_config.stay_on_initial_radio;
+    request->client_config().stay_on_selected_device = client_config.stay_on_selected_device;
+    request->client_config().selected_bands          = client_config.selected_bands;
+
+    int result = 0;
+    if (send_bml_cmdu(result, request->get_action_op()) != BML_RET_OK) {
+        LOG(ERROR) << "Send cACTION_BML_CLIENT_SET_CLIENT_REQUEST failed";
+        return (-BML_RET_OP_FAILED);
+    }
+
+    if (result != 0) {
+        LOG(ERROR) << "cACTION_BML_CLIENT_SET_CLIENT_REQUEST returned error code:" << result;
+        return result;
+    }
+
+    return BML_RET_OK;
+}
+
+int bml_internal::client_get_client(const sMacAddr &sta_mac, BML_CLIENT &client)
```

Comment:
Why you go back and forth from pointer to reference?   
`client` start as a pointer `int bml_client_get_client(BML_CTX ctx, const char *sta_mac, struct BML_CLIENT *client)`, then you change it to reference from here `pBML->client_get_client(tlvf::mac_from_string(std::string(sta_mac)), *client);` and back to pointer in `m_client = &client;`
```suggestion
int bml_internal::client_get_client(const sMacAddr &sta_mac, BML_CLIENT *client)
```

Reply:
Changed